### PR TITLE
Improve reporting keyboard layout changes

### LINKED ISSUE #7383: nvda announces keyboard and input language in weird places
Very hard to reproduce, and not expected unless I press Alt and shift.
It happens many times when I sign in. It announces my language and my keyboard input method in unexpected places.
It sometimes happens when I start some programs and the response takes a while.

### LINKED ISSUE #8419: NVDA fails to report system input language switching with Alt+Shift
### Steps to reproduce:
1. Update NVDA to the latest "next" snapshot (next-15324,255cb03f).
2. Try to switch between input languages with Alt+Shift. My input languages are 2: English U.S. and Bulgarian Bulgaria.

### Expected behavior:
NVDA should report the input language change.

### Actual behavior:
NVDA only plays its error sound.

### System configuration:
- Operating system: Windows 10 Pro version 1803 (build 17134.112), 64-bit, in Bulgarian with all locale settings set to "Bulgarian".
- NVDA version: next-15324,255cb03f.
- Processor: Intel Core i5-4670 at 3.40GHz.
- RAM Memory: 8.00GB.
- Graphics: MSI GeForce GTX-1050TI Gaming X 4G, 4096MB dedicated memory + 4065MB shared system memory, desktop resolution set to 1920x1080.
- Sound Card: Realtek ALC887 at Intel Lynx Point PCH - High Definition Audio Controller.

### Other questions:
I'm attaching the NVDA log with the logging level set to "Debug".
[NVDA_Log_2018-06-19.log](https://github.com/nvaccess/nvda/files/2114101/NVDA_Log_2018-06-19.log)

### PR DESCRIPTION
**Pull request by dkager**
### Link to issue number:
Fixes #7383. Fixes #8419.

### Summary of the issue:
If only one keyboard layout is available, NVDA still reports keyboard layout changes. This occurs most frequently when opening new windows. It appears that these notifications come from (de)activating a TSF profile. The chanes should have been ignored. A logic error caused them to be reported anyway.

The message would alternate between one with and one without the language name. For example:
- English (United States) - United States-International
- United States-International
- English (United States) - United States-International
- United States-International
- etc

### Description of how this pull request fixes the issue:
This  PR fixes the logic error. It also overhauls the code to return early if the old and new layouts are identical. NVDA now queries the active layout when it starts instead of comparing the first layout change to `None` and then always reporting it. Finally, support for XP is removed from the TSF C++ code as requested by @leonardder.

### Testing performed:
Tested by using the new code for 15-30 minutes and looking at all the notifications NVDA got and what it did with them. I used one layout. Extensive testing by people who actually use multiple layouts is still needed.

### Known issues with pull request:
Even if there is only one layout and NVDA would therefore never report chanes, it still listens for them. This could be avoided by checking how many layouts are installed when NVDA starts. However, this means that NVDA has to be restarted after adding or removing a keyboard layout. I think that the small overhead of just listening to layout changes is the better choice.

### Change log entry:
Bug fixes:
- NVDA no longer announces redundant keyboard layout changes. (#7383)

### CODE REVIEW

File: source/NVDAHelper.py [hunk 1]
```diff
@@ -1,3 +1,9 @@
+#NVDAHelper.py
+#A part of NonVisual Desktop Access (NVDA)
+#Copyright (C) 2017-2018 NV Access Limited, Peter Vagner, Davy Kager
```
Comment (leonardder) [inline source/NVDAHelper.py]: Thanks for adding the header, but I don't think 2017 is a very accurate guess here. Could you do a quick search with git blame and change this accordingly?

Comment (dkager) [reply source/NVDAHelper.py]: I already know it's 2008, not sure why that got lost.

**Review by LeonarddeR — DISMISSED** [review]
It is indeed hard to predict whether this works 100% as you intend it to work. Therefore, let's give it some time in next to let it prove itself.

**Comment by ehollig** [comment]
This sounds as if it fixes #5545 as well.

**Review by michaelDCurran — APPROVED** [review]

**Review by michaelDCurran — CHANGES_REQUESTED** [review]
I have multiple layouts installed including Chinese, Japanese and English US.  When I switch to say the japanese layout, the language is no longer announced which means I do not know it is Japanese. All it says when I switch is "Microsoft IME". This is similar for Chinese: All it says is "Microsoft Bopomofo". It used to say "Chinese Traditional (Taiwan) - Microsoft Bopomofo".

**Comment by dkager** [comment]
@michaelDCurran Sorry, this is fixed now.

**Review by michaelDCurran — DISMISSED** [review]

**Comment by gdata1** [comment]
Hello. After this update, the NVDA does not tell you to switch the layout when you switch error occurs.

**Comment by gdata1** [comment]
INFO - __main__ (11:31:16.726):
Starting NVDA
INFO - core.main (11:31:16.835):
Config dir: C:\Users\nikit\AppData\Roaming\nvda
INFO - config.ConfigManager._loadConfig (11:31:16.835):
Loading config: C:\Users\nikit\AppData\Roaming\nvda\nvda.ini
INFO - core.main (11:31:17.092):
NVDA version next-15324,255cb03f
INFO - core.main (11:31:17.092):
Using Windows version 10.0.17692 workstation
INFO - core.main (11:31:17.092):
Using Python version 2.7.15 (v2.7.15:ca079a3ea3, Apr 30 2018, 16:22:17) [MSC v.1500 32 bit (Intel)]
INFO - core.main (11:31:17.094):
Using comtypes version 1.1.3
INFO - external:synthDrivers.RHVoice.SynthDriver.__init__ (11:31:17.335):
Using RHVoice version 0.7.0
INFO - synthDriverHandler.setSynth (11:31:17.338):
Loaded synthDriver RHVoice
INFO - core.main (11:31:17.338):
Using wx version 4.0.1 msw (phoenix)
INFO - brailleInput.initialize (11:31:17.344):
Braille input initialized
INFO - braille.initialize (11:31:17.345):
Using liblouis version 3.6.0
INFO - braille.BrailleHandler.setDisplayByName (11:31:17.348):
Loaded braille display driver noBraille, current display has 0 cells.
WARNING - core.main (11:31:17.398):
Java Access Bridge not available
INFO - _UIAHandler.UIAHandler.MTAThreadFunc (11:31:17.404):
UIAutomation: IUIAutomation5
INFO - core.main (11:31:17.687):
NVDA initialized
ERROR - unhandled exception (11:31:25.529):
Traceback (most recent call last):
  File "_ctypes/callbacks.c", line 315, in 'calling callback function'
  File "NVDAHelper.pyc", line 368, in nvdaControllerInternal_inputLangChangeNotify
UnicodeEncodeError: 'ascii' codec can't encode characters in position 0-9: ordinal not in range(128)
ERROR - unhandled exception (11:31:25.546):
Traceback (most recent call last):
  File "_ctypes/callbacks.c", line 315, in 'calling callback function'
  File "NVDAHelper.pyc", line 368, in nvdaControllerInternal_inputLangChangeNotify
UnicodeEncodeError: 'ascii' codec can't encode characters in position 0-9: ordinal not in range(128)
ERROR - unhandled exception (11:31:27.565):
Traceback (most recent call last):
  File "_ctypes/callbacks.c", line 315, in 'calling callback function'
  File "NVDAHelper.pyc", line 368, in nvdaControllerInternal_inputLangChangeNotify
UnicodeEncodeError: 'ascii' codec can't encode characters in position 0-9: ordinal not in range(128)
ERROR - unhandled exception (11:31:27.585):
Traceback (most recent call last):
  File "_ctypes/callbacks.c", line 315, in 'calling callback function'
  File "NVDAHelper.pyc", line 368, in nvdaControllerInternal_inputLangChangeNotify
UnicodeEncodeError: 'ascii' codec can't encode characters in position 0-9: ordinal not in range(128)

**Comment by dkager** [comment]
@gdata1 The most recent commit should fix this. Please give that build a try or wait until it lands in next.

**Review by michaelDCurran — APPROVED** [review]

**Pull request merged**